### PR TITLE
(WIP) Fixes #902 added a way to sent multipart/form-data requests via play-ws

### LINKED ISSUE #902: Provide a post() method on WS for POSTing multipart/form-data
There is actually no way to post a multipart/form-data, without encoding manually the body (and this is tricky!)

It's sad to not have a WS method which allow to call a controller like the "[File upload example](http://www.playframework.com/documentation/2.1.0/ScalaFileUpload)"

### PR DESCRIPTION
**Pull request by schmitch**
Hello here is a PR for #902, actually it misses the following things:

- [x] Tests
- [x] Squashed Commits
- [x] Java Code

Actually before providing a Documentation I wanted to grab some feedback if that is clever enough.
Also I'm not sure where to put tests for that behavior since there aren't many tests for WS yet.

**Edit:** 
Actually what I thought is if I should provide custom interfaces for Part, however then I thought that actually implementing them on top would actually double the work here, however It could be good to add them later, so that we could send JsValue and XML without additional work. However for the moment I found that satisfying, since there aren't many frameworks that actually parse anything else than StringParts and FileParts, actually even Play only knows how to handle Content-Disposition=form-data. 

**Edit2:** I added documentation, I'm still unsure about additional tests, are they required? And when, where could I put them? Would've put them into ``play.libs.ws.WSSpec`` and ``play.api.libs.ws.WSRequestSpec``

### CODE REVIEW

**Comment by gmethvin** [comment]
Awesome. @wsargent should probably review.

**Comment by wsargent** [comment]
Hi @schmitch -- so the documentation is good on its own terms -- the issue is that we'd like to be able to send `WSPart` without having to import org.asynchttpclient classes directly, so there needs to be a shell on top for both Scala and Java APIs.

The other thing is that we'd want to attach a Streams API if possible to form upload, and the FeedableBodyGenerator (at least last time I looked) didn't implement backpressure.

**Comment by schmitch** [comment]
@wsargent are case classes which are wrapper around Part okay? And what About java should it have his own wrapper? Actually i'm unsre about the streams api. If i would create it i would've either Need to materialize it before sending or provide a custom multipart body generator. Actualiy file upload materializes before sending as i remember so that wouldn't backpressure. I thought about the streams api, bunt implementing that with asynchttpclient is way harder than just using Akka-Http .

Actually there is a way to insert a Streambody however i would Need to create a custom multipart Generator on top of the streams api.

**Comment by schmitch** [comment]
@wsargent I added wrapper classes for scala now, are something like that sufficant?

Also I further looked into the Streams API. Something that I came up recently is, that if you want a 100% streamable file upload API with mutlipart/form data you either just use akka-http or you pretty much need to build a lot from scratch.
The only thing to have a Stream Like API Akka-Streams to Ahc would be the setBody(Publisher<ByteBuffer>) method. which uses FeedableBodyGenerator.

However since the current Implementation of Ahc's Multipart/form data is actually a stringbuilder you would've need to reimplement nearly everything for that. to push a multipart to the setBody method.
The other way would be creating a SourcePart and materialize it into a ByteArrayPart before sending, however that wouldn't backpressure and I would call that 'cheating', cause of this issue's I actually thought about just creating a play-ws2 (internally) which would've been built on top of akka-http. However some people would be against that. Even that akka-http isn't as full featured as Ahc it's still is somewhat better in the underlying api. I mean bringing a stream api to multipart/form data in Ahc is way out of scope.

**TL/DR:** actually at the current stage of ahc I wouldn't implement a streams api on top of ahc's multipart/form data.

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala [hunk 1]
```diff
@@ -689,3 +734,91 @@ case class DefaultWSResponseHeaders(status: Int, headers: Map[String, Seq[String
  * Sign a WS call with OAuth.
  */
 trait WSSignatureCalculator
+
+/**
+ * A Part of multipart/form data request
+ */
+trait WSPart {
```
Comment (jroper) [inline framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala]: Should be sealed.

**Comment by jroper** [comment]
I think there should be only two types of `WSPart`:

1. `DataPart`.  This is a simple in memory part, equivalent to the MVC multipart form data API.  It would conceptually by equivalent to a text input field.  It would have a `name` and a `value`, both of type `String`.
2. `FilePart`.  This is conceptually equivalent to a file input field, and would have a `name`, `filename`, optional `contentType`, and most importantly, a `content` field with a type of `Source[ByteString, _]`.  Files are not the only things that you might want to send, in fact very commonly you will want to stream a request body from the server through to the client without it having to touch the disk.

Also, while a helper may exist to allow passing a `List[WSPart]`, the main API should allow passing a `Source[WSPart, _]`.

In practical terms, implementing all this actually means we don't need to make any fundamental changes to the WS API - since it can all be implemented using the `setBody(Source[ByteString, _])`, with an appropriate `Flow[WSPart, ByteString, _]` that converts the `Source[WSPart, _]` to the formatted body.  That said, some helper methods will need to be created since a boundary needs to be generated, which would mean setting the content type to have that boundary in it.

It also means that we don't use the underlying AHC multipart form data support, instead, we implement the formatting of the multipart body using Akka streams.  The advantage here is that the same code that formats the WS `multipart/form-data` bodies can also be used when implementing tests for actions that accept `multipart/form-data` bodies.

Finally, I think ideally I'd rather not introduce a new API for `multipart/form-data` when one already exists in Play - in `play.api.mvc.MultipartFormData.Part` for Scala, and `Http.MultipartFormData` for Java.  These APIs may not be quite ready for consumption by the WS API, but it shouldn't be hard to modify them so they are.  This would also aid in the dual purpose of using the same code for WS formatting as is used in formatting the bodies in testing.  In general, we should be aiming for more unification between the `play.api.mvc` APIs and the `play.api.ws` APIs, and so introducing a new feature like this is a perfect opportunity to do so.

**Comment by schmitch** [comment]
I've seen that coming, however I was a little bit too scared to use things in ``play.api.mvc`` and to create a custom multipart/form-data formatter. However you are completly right that this will help on testing stuff, actually I've maybe already hit a multipart/form-data parsing bug on 2.5-RC1 but I'm not sure if thats my fault. 

However one question, @jroper should I implement a writeable on top of that or could I keep my def ``post(List[WSPart])`` and change the ``post(WSPart)`` to  ``post(Source[WSPart, _]``.

I will try my best to implement it like you suggested. But I will take some time. (maybe a week)

**Comment by schmitch** [comment]
@jroper I added a new way of creating ws multipart requests under scala it's not merge ready (not near) could you take a look if that would fit better? It makes use of ``DataPart`` and ``FilePart[Source[ByteString, _]]`` actually it's just a quick prototype.

**Comment by mkurz** [comment]
@schmitch A bit off-topic - anyway:
I haven't really looked into this, but would a fix for #2289 and #2125 (which look like duplicates) be similiar to this PR? What do you think?

**Comment by jroper** [comment]
@schmitch That's exactly what I was thinking - and starting with the code from akka-http is good, that's exactly what I did for the multipart/form-data parser.

I imagine we should be able to put this into a point release of Play 2.5 if it doesn't make 2.5.0.

A few things:

* `BodyPartHelper` should be moved to Play core, it is going to be useful to more than just ws.  One possibility is to put it in `play.core.parsers.Multipart`, though it's not really a parser, so maybe `play.core.formatters.Multipart`.  I don't have a strong opinion here.
* `Formatters` should be slimmed down (remove anything we don't need) and put into the `Multipart` object as a provite class.  If this API is going to be useful elsewhere, then we should force whoever wants to use it make a decision then of where it should live, rather than just make it available for adhoc usage.
* The `transform` method should take a `Source[Part]`, rather than a `Seq[Part]`.  The first thing the method does is convert the `Seq` to a `Source` anyway, `Source` should be the lowest level API, and helpers that take a `Seq` can be easily built on top of that.
* Similar to the previous point, the `BodyPartHelper` could provide a method that looks like this:

```scala
def format(boundary: String /* and maybe other params, charset etc */): Flow[Part[...], ByteString, NotUsed]
```

An API like this would be suitable to make public and provides the highest level of flexibility.  With that, the implementation of `transform` (which now takes a `Source`) just needs to do `body.via(flow(..))` to convert the source.

**Comment by schmitch** [comment]
@jroper I did what you wrote, hopefully this takle's everything you told me. 
Actually I hope I could do similar on the Java side.

@mkurz Actually they are related but unfortunately my *quick* look at those issue's prevent me from saying that this PR would fix those. Actually my PR will add an easy way to add a MultipartBody to everything that has a Source[ByteString] interface, which actually RequestBuilder has not / not yet, I'm not sure if there will be one. However when you have a Source[ByteString] you could easily materialize that into a Future[ByteString] so that these people could actually add bodies with .asRaw, not sure if I should provide some helper for that on the Java part, I actually start tackling that as soon as the Scala implementation is fine.

File: framework/src/play/src/main/scala/play/core/formatters/Multipart.scala [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        def completePartFormatting(): Source[ByteString, Any] = bodyPart match {
+          case MultipartFormData.DataPart(_, data) => Source.single((f ~~ ByteString(data)).get)
+          case MultipartFormData.FilePart(_, _, _, ref) => bodyPartChunks(ref)
+          case _ => throw new UnsupportedOperationException()
+        }
+
+        renderBoundary(f, boundary, suppressInitialCrLf = !firstBoundaryRendered)
+        firstBoundaryRendered = true
+
+        val (key, filename, contentType) = bodyPart match {
```
Comment (schmitch) [inline framework/src/play/src/main/scala/play/core/formatters/Multipart.scala]: should also ```throw new UnsupportedOperationException()`` on ```case _ =>```

**Comment by schmitch** [comment]
Rebased and added Copyright Header, could somebody look at comment and help me out with the Java API?

**Comment by schmitch** [comment]
I squashed it and fixed the Java API, however there is a IntelliJ error in the MultipartFormatter, however compiling, etc works fine.

File: documentation/manual/working/javaGuide/main/ws/code/javaguide/ws/JavaWS.java [hunk 3]
```diff
@@ -114,6 +115,11 @@ public void requestExamples() {
             ws.url(url).post(json);
             // #ws-post-json
 
+            // #ws-post-multipart
+            ws.url(url).post(Source.single(new Http.MultipartFormData.DataPart("hello", "world")))
```
Comment (marcospereira) [inline documentation/manual/working/javaGuide/main/ws/code/javaguide/ws/JavaWS.java]: Build reports an error here due to the lack of a ";" at the end of line:

https://travis-ci.org/playframework/playframework/jobs/113388976#L580

Comment (schmitch) [inline documentation/manual/working/javaGuide/main/ws/code/javaguide/ws/JavaWS.java]: Thanks, I also had a mistake in the documentation itself where i written: 
@[ws-post-multipart(code/javaguide/ws/JavaWS.java)

File: framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSRequest.java [hunk 4]
```diff
@@ -331,6 +334,15 @@ String getVirtualHost() {
         return execute();
     }
 
+    public CompletionStage<WSResponse> patch(Source<Http.MultipartFormData.Part<Source<ByteString, ?>>, ?> body) {
+        String boundary = MultipartFormatter.randomBoundary();
```
Comment (marcospereira) [inline framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSRequest.java]: We can avoid some repetition here by rewrite these methods to something like this:

```java
public CompletionStage<WSResponse> patch(Source<Http.MultipartFormData.Part<Source<ByteString, ?>>, ?> body){
    setMethod("PATCH");
    return executeWithBody(body);
}

private CompletionStage<WSResponse> executeWithBody(Source<Http.MultipartFormData.Part<Source<ByteString, ?>>, ?> body) {
    String boundary = MultipartFormatter.randomBoundary();
    Source<ByteString, ?> innerBody = MultipartFormatter.transform(body, boundary);

    setBody(innerBody);
    setHeader("Content-Type", MultipartFormatter.boundaryToContentType(boundary));
    return execute();
}
```

File: framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala [hunk 5]
```diff
@@ -474,6 +476,15 @@ trait WSRequest {
   def patch(body: File): Future[WSResponse] = withMethod("PATCH").withBody(FileBody(body)).execute()
 
   /**
+   * Perform a POST on the request asynchronously.
```
Comment (marcospereira) [inline framework/src/play-ws/src/main/scala/play/api/libs/ws/WS.scala]: "Perform a ~~POST~~ PATCH on the request asynchronously."

File: framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSRequest.java (hunk 4 above)
Comment (schmitch) [inline framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSRequest.java]: Is it valid to do the same on the scala version?
I.e. adding the following to the WSRequest trait:

```scala
  private[libs] def withMultipartBody(body: Source[MultipartFormData.Part[Source[ByteString, _]], _]): WSRequest = {
    val boundary = Multipart.randomBoundary()
    val contentType = s"multipart/form-data; boundary=$boundary"
    withBody(StreamedBody(Multipart.transform(body, boundary))).withHeaders("Content-Type" -> contentType)
  }
```

Comment (marcospereira) [inline framework/src/play-java-ws/src/main/java/play/libs/ws/ahc/AhcWSRequest.java]: Sounds valid to me. ;-)

**Comment by mkurz** [comment]
What's the state of this PR? Is it ready to merge?

**Comment by schmitch** [comment]
Actually it's ready to review.

File: documentation/manual/working/javaGuide/main/ws/JavaWS.md [hunk 6]
```diff
@@ -79,6 +79,12 @@ The easiest way to post JSON data is to use the [[JSON library|JavaJsonActions]]
 
 @[ws-post-json](code/javaguide/ws/JavaWS.java)
 
+### Submitting multipart/form data
+
+The easiest way to post multipart/form data is to use a Source<Http.MultipartFormData<Source<ByteString>, ?>, ?>
```
Comment (gmethvin) [inline documentation/manual/working/javaGuide/main/ws/JavaWS.md]: You can use backticks in markdown to render the code in monospace typeface: `` `Source<Http.MultipartFormData<Source<ByteString>, ?>, ?>` ``

File: framework/src/play/src/main/java/play/mvc/Http.java [hunk 7]
```diff
@@ -1450,6 +1455,35 @@ public A getFile() {
 
         }
 
+        public static class DataPart implements Part<Source<ByteString, ?>> {
+            final String key;
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/mvc/Http.java]: private

File: framework/src/play/src/main/java/play/mvc/MultipartFormatter.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+
+public class MultipartFormatter {
+
+    public static String randomBoundary() {
+        return Multipart.randomBoundary(18, ThreadLocalRandom.current());
+    }
+
+    public static String boundaryToContentType(String boundary) {
+        return new StringBuilder().append("multipart/form-data; boundary=").append(boundary).toString();
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/mvc/MultipartFormatter.java]: Simple string concatenation should be fine here.

File: framework/src/play/src/main/java/play/mvc/MultipartFormatter.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public static Source<ByteString, ?> transform(Source<? super Http.MultipartFormData.Part<Source<ByteString, ?>>, ?> parts, String boundary) {
```
Comment (gmethvin) [inline framework/src/play/src/main/java/play/mvc/MultipartFormatter.java]: Shouldn't this use `? extends Http.MultipartFormData.Part`? That should avoid the need for casts below.

File: framework/src/play/src/main/scala/play/core/formatters/Multipart.scala [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      builder += char.toByte
+      this
+    }
+
+  }
+
+  private def streamed(boundary: String,
+    nioCharset: Charset,
+    chunkSize: Int): PushPullStage[MultipartFormData.Part[Source[ByteString, _]], Source[ByteString, Any]] =
+    new PushPullStage[MultipartFormData.Part[Source[ByteString, _]], Source[ByteString, Any]] {
```
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/core/formatters/Multipart.scala]: I wonder if we can easily use `GraphStage` here rather than `PushPullStage`? Though `play.core.parsers.Multipart` still uses `PushPullStage` so we could probably migrate both at the same time.

Comment (schmitch) [inline framework/src/play/src/main/scala/play/core/formatters/Multipart.scala]: Actually I used `PushPullStage` since akka-streams used it and I think the API is somewhat easier to use than `GraphStage`. (https://github.com/akka/akka/blob/0d8dd3c7114ed5258208aaed3575460a056876bb/akka-http-core/src/main/scala/akka/http/impl/engine/rendering/BodyPartRenderer.scala#L32) Actually there are two ways of migrating, either I do it know but the implementation will propably diverege from akka-streams or we keep an eye on the `BodyPartRenderer` and change it as soon as they changed it. I think the latter is better since Bugs could be easier tracked since they would most likely appear in both projects or am I wrong?

File: framework/src/play/src/main/java/play/mvc/Http.java (hunk 7 above)
Comment (schmitch) [inline framework/src/play/src/main/java/play/mvc/Http.java]: Do you mean the `final String key` and the `final String value` ? Does that change anything since they are final, too. I didn't made them private since FilePart doesn't have `private` on their properties, too.

File: framework/src/play/src/main/java/play/mvc/MultipartFormatter.java (hunk 9 above)
Comment (schmitch) [inline framework/src/play/src/main/java/play/mvc/MultipartFormatter.java]: extends or super doesn't make any real difference, since I need to convert between `Http.MultipartFormData.DataPart` to `MultipartFormData.DataPart` and `Http.MultipartFormData.FilePart` to `MultipartFormData.FilePart`  would've been way easier if Scala and Java would use somewhat the same API. (Actually I need to fight against type inference here)

File: framework/src/play/src/main/java/play/mvc/Http.java (hunk 7 above)
Comment (gmethvin) [inline framework/src/play/src/main/java/play/mvc/Http.java]: Leaving off `private` just means they available within the current package. Since you have the accessor methods it is probably a better API to make them private.

File: framework/src/play/src/main/java/play/mvc/MultipartFormatter.java (hunk 9 above)
Comment (gmethvin) [inline framework/src/play/src/main/java/play/mvc/MultipartFormatter.java]: Actually I was wrong about the cast. Anyway not a big deal since this is sort of internal code.

File: framework/src/play/src/main/scala/play/core/formatters/Multipart.scala (hunk 10 above)
Comment (gmethvin) [inline framework/src/play/src/main/scala/play/core/formatters/Multipart.scala]: Yeah, Akka is still using it too for multipart, so we could probably just wait until they remove it and force themselves to upgrade, then  it would be easier to migrate by looking at their new implementation.

**Comment by mkurz** [comment]
Build is green. @gmethvin Maybe we could get this in 2.5.1?

**Comment by gmethvin** [comment]
lgtm but I'd like to see a few tests. You could add a few to `play.it.libs.WSSpec`.

**Comment by schmitch** [comment]
@gmethvin I added some tests.
Actually it was good that I made some tests since the Java implementation needed a small change.

File: documentation/manual/working/scalaGuide/main/ws/ScalaWS.md [hunk 11]
```diff
@@ -216,7 +222,6 @@ You can get access to the underlying [AsyncHttpClient](http://static.javadoc.io/
 
 This is important in a couple of cases.  WS has a couple of limitations that require access to the underlying client:
 
-* `WS` does not support multi part form upload directly.  You can use the underlying client with [RequestBuilder.addBodyPart](http://static.javadoc.io/org.asynchttpclient/async-http-client/2.0.0-RC12/org/asynchttpclient/RequestBuilderBase.html#addBodyPart-org.asynchttpclient.request.body.multipart.Part-).
```
Comment (mkurz) [inline documentation/manual/working/scalaGuide/main/ws/ScalaWS.md]: @schmitch Shouldn't this be deleted in the `JavaWS.md` as well?

**Comment by gmethvin** [comment]
LGTM, besides the issue mentioned by @mkurz.

**Comment by schmitch** [comment]
I removed the part inside `JavaWS.md`.

**Comment by mkurz** [comment]
@gmethvin Build is green

**Comment by mkurz** [comment]
Backport: #5887

**Comment by mkurz** [comment]
@schmitch Thanks a lot for your time and providing this fix! I am quite happy to finally have this working out of the box.
Can I ask you for a favor? As you are familiar with the multipart stuff, API, RFC's, etc. now could you have a look at #2289 and #2125 (which look like they are duplicates anyway) and maybe provide a fix for them as well if you find the time? It would be really great to get this long standing issues fixed finally as well.

**Comment by schmitch** [comment]
@mkurz for #2289 a PR is planned soon (probably I take a look on sunday or on easter since I have some holidays there), and withFile is different, that means sending a file directly. However I could tackle that aswell.
Also I'm mostly fine with multipart stuff, but some RFC's are difficult to read, however there are great people who even helped me on this like wsargent.

**Comment by mkurz** [comment]
@schmitch Great to hear! Looking forward to it :smile: Thanks a lot - I really appreciate it!

**Comment by wsargent** [comment]
So... can we close https://github.com/playframework/playframework/issues/2125 now?

**Comment by schmitch** [comment]
No #2125 is *only* sending a file, however In the upcoming week. The question is only, do we really need a `withFile(java.io.File)`? I mean with bodyRaw you could already sent files when you read them to a `Array[Byte]` or `ByteString` and Streaming the file isn't possible since the `route()` functions of `testing` don't provide a streaming interface.

**Comment by marcospereira** [comment]
What would be the difference between a `withFile` method and using a `FilePart`?

**Comment by schmitch** [comment]
From my point of understanding, `withFile` would send a file directly without encoding, while the `FilePart` that goes through `withMultipart` will actually be a multipart body, but that's what I said `withFile` is useless, since you could just use `bodyRaw` after you have your file in a `Array[Byte] ` or `ByteString`, since the RequestBuilder would need to convert it internally anyway.

**Comment by marcospereira** [comment]
Thanks, @schmitch. I also don't see the point of having `withFile` method. `bodyRaw` even looks like a better name to me.
